Add basic functionality of leapp report analyzer

Wecros opened the pull request:
Introduce script to analyze and group multiple leapp report JSON files into a single JSON output and an HTML tables output.

The scripts takes a list of input files, merges their data, and groups the hostnames by severity.

Output is sorted by severity.

Add support for checking the 'groups' field if the 'flags' field is not present in the leapp report entry. So we can parse multiple versions of leapp report schemas.

The script seems to work for both python2.7 and 3+ versions.

Usage: leapp-report-analyzer.py [-h] -o OUTPUT_JSON -t OUTPUT_HTML input_file [input_file ...]
```
leapp-report-analyzer leapp-report1.json leapp-report2.json -o output.json -t output.html
```

Wecros committed:
Add basic functionality of leapp report analyzer

Introduce script to analyze and group multiple leapp report JSON files
into a single JSON output and an HTML tables output.

The scripts takes a list of input files, merges their data, and groups
the hostnames by severity. Output is sorted by severity.

Add support for checking the 'groups' field if the 'flags' field is not
present in the leapp report entry. So we can parse multiple versions of
leapp report schemas.

The script seems to work for both python2.7 and 3+ versions.

andywaltlova commented:
Ok, so after discussion, just leaving a comment to include README.md with description (+license), basic usage. Just make it representative as standalone tool.

Wecros committed:
Update readme to include vital information

The idea is to enable this repository to be used as a standalone tool
for customers analyzing their systems.

Wecros commented:
@andywaltlova Updated the readme.

andywaltlova commented on README.md:
```diff
@@ -1 +1,43 @@
+# Leapp Report Analyzer
+
+Leapp Report Analyzer is a command-line tool for analyzing multiple Leapp reports and grouping systems by severity and entries.
+The tool processes JSON files containing Leapp reports and generates a JSON file and an HTML table displaying the results.
```
I would link here the leapp report schema, or appropriate part of leapp documentation.

In first sentence then link leapp repo/docs.

Wecros replied:
Updated with links to leapp repo and docs. Unsure of usefulness to customers, but it's imo good to give context of the repo here.

andywaltlova commented on README.md:
```diff
@@ -1 +1,43 @@
+
+You can use the Leapp Report Analyzer with either individual Leapp report JSON files or a directory containing multiple JSON files. The following arguments are required:
+
+- `-f / --input-files`: One or more Leapp report JSON files to be analyzed (e.g., `-f report1.json report2.json`).
+- `-d / --input-directory`: A directory containing Leapp report JSON files to be analyzed (e.g., `-d reports_directory`).
```
Both, or only one of them is required? based on code they are exclusive, so -f or -d

Wecros replied:
updated

andywaltlova commented on README.md:
```diff
@@ -1 +1,43 @@
+- Python 2.7
+- Python 3.3
+
+### Usage
```
you could just paste here `python3 leapp-report-analyzer.py -h` or `python leapp-report-analyzer.py -h`

Wecros replied:
updated

andywaltlova reviewed (changes requested):
Works as expected, I won't comment the format of output - can be adjusted to customer needs afterwards, looks like good starting point :+1:

pirat89 commented on README.md:
```diff
@@ -1 +1,60 @@
+Script works on both Python2 and Python3, tested with:
+
+- Python 2.7
+- Python 3.3
```
why python 3.3 which is not supported anyhwere? I would expect 3.6 or 3.9+ to be honest

andywaltlova replied:
code doesn't have any version specific syntax, so maybe it would just be best to write that it works on all python 2 a 3 supported versions

pirat89 replied:
py2 is not supported anymore officially by Python community so it could be a little bit confusing. I would rather write something like:
```
- Python 2.7
- Python 3.6+
```

As there is not used effectively any Python 3.[0-5] nowadays.

Wecros replied:
I understand. The point is that system administrators can have various infrastructure at their company, so staying on the safe side and listing earlier versions looks okay to me, it doesn't even have to be run on RHEL machines (though it can...).

I tested it now on python3.2 and newer versions and seems to work fine. I think it will break on python3.1 since `argparse` is supported since Python3.2 and Python2.7, so it should also break on python2.6 (both archaic 😄).

Based on this info I would change the readme as follows (edited in commit):
"""
Script works on both Python2 and Python3, versions:

- Python 2.7
- Python 3.2+

"""

andywaltlova replied:
"""
Python 2.7
Python 3.2+
"""
Alright above is fine with me

Wecros committed:
Update script to include remediations and keys

Remediations are now included both in json and in html. Depending on
type "hint", "command", "playbook" they are displayed differnetly in the
HTML.

Entries are now grouped by key. Some entry titles are repeated, so using
unique keys (which group same remediations together) is effective for
most effectively grouped remediations.

Legend table has also been added to the html file with listed severities
levels from most crucial to least crucial with description.

Wecros commented:
I updated the script and html output with remediations, feel free to review @andywaltlova

andywaltlova commented on leapp-report-analyzer.py:
```diff
@@ -0,0 +1,199 @@
+    return escape(text, quote=True)
+
+
+def main():
```
My preference would be to put main function to the bottom, it feels more natural to me to have necessary functions first and then function that uses them (in this case main).

Wecros replied:
I disagree. I like top-down approach for structuring code. Most important first so that the code unwraps in a natural way to the bottom, not going down and than back up, each to their own I guess.

I only placed some helper functions at the top which are not really that much related to other functions.

Is this okay with you?

andywaltlova replied:
Alright.. well I can live with it if I have to :D .. seems confusing though that the ` _create_grouped_data_structure()` would then be above main, it's kinda important to other functions but not used in main.. :D either put the main at top or bottom and everything else under or above it.

Now it doesn't even work as you are describing it, if I want to naturally unwrap it then I want main -> functions in main -> functions used in functions in main -> ...

Wecros replied:
yeah I would put the functions into some `helper.py` file since they seem to have different purpose than other functions (but we want to keep it all in one file), but I see your point about consistency, so will put main as top.

Added.

andywaltlova replied:
alright, nice

andywaltlova commented on leapp-report-analyzer.py:
```diff
@@ -0,0 +1,199 @@
+    else:
+        input_files = args.input_files
+
+    grouped_data_list = [parse_leapp_report(input_file) for input_file in input_files]
+    merged_grouped_data = merge_grouped_data(grouped_data_list)
+    ordered_merged_data = order_grouped_data(merged_grouped_data)
+    write_grouped_data_to_json(ordered_merged_data, args.output_json)
+    html_tables = generate_html_tables(ordered_merged_data)
+    write_html_tables_to_file(html_tables, args.output_html)
```
Can we have some basic logging? At least some basic info what is happening during the execution.

Wecros replied:
Added

andywaltlova commented on leapp-report-analyzer.py:
```diff
@@ -0,0 +1,199 @@
+    )
+
+    output_group = parser.add_argument_group("output", "Output options, both required")
+    output_group.add_argument("-o", "--output-json", required=True, help="Output JSON file to store the grouped data")
+    output_group.add_argument(
+        "-t", "--output-html", required=True, help="Output HTML file to store the tables with grouped data"
+    )
```
seems to me that both of these could be optional, if they are not specified why not just use `'leapp_report_analysis_output.html'` and `'leapp_report_analysis_output.html'` ? maybe add timestamp

Wecros replied:
I do not think timestamp really gives any additional information here that much, don't want to make it complicated without much benefit. I put the timestamp info into seperate issue as discussed in other comment though.

Added the default args.

andywaltlova commented on leapp-report-analyzer.py:
I am not sure how (and by who) will the output be used in the end, but I think somewhere should be some timestamp when the report was generated - property/filename.. (maybe even what reports were used as input ? - that maybe too much and unnecessary, but it crossed my mind)

Wecros replied:
added in Jira tracker

Wecros replied:
Oh, I meant the sanity tests and linting. The other ideas are good, but I am unsure in what way to implement them rn, so adding them to jira tracker again.

andywaltlova commented:
And please create  issue(s) in the jira story for any subsequent tasks that should be done but are now omitted - like sanity tests, lint.

Wecros committed:
Add basic logging

Useful for users of script to know what is happening and if something
failed and where.

Wecros committed:
Convert output arguments to optional

Use default file names if output args are not present.

Add information about default output names to readme.

Wecros committed:
Restructure code to follow top-bottom structure
